File: src/propagation.rs
use crate::wit::v2::observe::get_parent_span_context;
use opentelemetry::{
    trace::{
        SpanContext as OtelSpanContext, SpanId, TraceContextExt, TraceFlags, TraceId, TraceState,
    },
    Context, ContextGuard,
};
use std::str::FromStr;

pub fn extract_trace_context() -> anyhow::Result<ContextGuard> {
    let other_sc = get_parent_span_context();

    let sc = OtelSpanContext::new(
        TraceId::from_hex(&other_sc.trace_id)?,
        SpanId::from_hex(&other_sc.span_id)?,
        TraceFlags::SAMPLED,
        other_sc.is_remote,
        TraceState::from_str(&other_sc.trace_state)?,
    );

    Ok(Context::current().with_remote_span_context(sc).attach())
}

File: go/propagation.go
package wasi_otel

import (
	"context"

	wasiTrace "github.com/calebschoepp/opentelemetry-wasi/internal/wasi/otel/tracing"
	"go.opentelemetry.io/otel/trace"
)

type WasiPropagator interface {
	Extract(cx context.Context) context.Context
}

type TraceContextPropagator struct{}

func NewTraceContextPropagator() *TraceContextPropagator {
	return &TraceContextPropagator{}
}

func DefaultTraceContextPropagator() *TraceContextPropagator {
	return NewTraceContextPropagator()
}

func (t TraceContextPropagator) Extract(cx context.Context) context.Context {
	// Retrieving span context from the wasm host
	hostCx := wasiTrace.OuterSpanContext()

	// Converting wasm host TraceState type to otel TraceState type
	var traceState trace.TraceState
	for _, entry := range hostCx.TraceState.Slice() {
		ts, err := traceState.Insert(entry[0], entry[1])
		if err != nil {
			// TODO: not sure how to handle this error
		}
		traceState = ts
	}

	cfg := trace.SpanContextConfig{
		TraceID:    trace.TraceID([]byte(hostCx.TraceID)),
		SpanID:     trace.SpanID([]byte(hostCx.SpanID)),
		TraceFlags: trace.TraceFlags(hostCx.TraceFlags),
		TraceState: traceState,
		Remote:     hostCx.IsRemote,
	}

	convertedCx := trace.NewSpanContext(cfg)

	return trace.ContextWithRemoteSpanContext(cx, convertedCx)
}
